gsplat/strategy.py

```diff
@@ -0,0 +1,552 @@
+
+        is_split = is_grad_high & ~is_small
+        is_split = torch.cat(
+            [
+                is_split,
+                # new GSs added by duplication will not be split
+                torch.zeros(n_dupli, dtype=torch.bool, device=device),
+            ]
+        )
+
+        n_split = is_split.sum().item()
+        if n_split > 0:
+            self._refine_split(is_split)
+        return n_dupli, n_split
+
+    @torch.no_grad()
+    def _prune_gs(self, step) -> int:
+        is_prune = torch.sigmoid(self._params["opacities"]) < self._prune_opa
+        if step > self._reset_every:
+            # The official code also implements sreen-size pruning but
+            # it's actually not being used due to a bug:
+            # https://github.com/graphdeco-inria/gaussian-splatting/issues/123
+            is_too_big = (
+                torch.exp(self._params["scales"]).max(dim=-1).values
+                > self._prune_scale3d * self._scene_scale
+            )
+            is_prune = is_prune | is_too_big
+
+        n_prune = is_prune.sum().item()
+        if n_prune > 0:
+            self._refine_keep(~is_prune)
+        return n_prune
+
+    def _update_state(self, info: Dict[str, Any], packed: bool = False):
+        for key in ["means2d", "width", "height", "n_cameras", "radii", "gaussian_ids"]:
+            assert key in info, f"{key} is required but missing."
+
+        # normalize grads to [-1, 1] screen space
+        if self._absgrad:
+            grads = info["means2d"].absgrad.clone()
+        else:
+            grads = info["means2d"].grad.clone()
+        grads[..., 0] *= info["width"] / 2.0 * info["n_cameras"]
+        grads[..., 1] *= info["height"] / 2.0 * info["n_cameras"]
+
+        # initialize states on the first run
+        n_gaussian = len(list(self._params.values())[0])
+        if self._state["grad2d"] is None:
+            self._state["grad2d"] = torch.zeros(n_gaussian, device=grads.device)
+        if self._state["count"] is None:
+            self._state["count"] = torch.zeros(n_gaussian, device=grads.device)
+
+        # update the running state
+        if packed:
+            # grads is [nnz, 2]
+            gs_ids = info["gaussian_ids"]  # [nnz]
+            self._state["grad2d"].index_add_(0, gs_ids, grads.norm(dim=-1))
+            self._state["count"].index_add_(
+                0, gs_ids, torch.ones_like(gs_ids, dtype=torch.float32)
+            )
+        else:
+            # grads is [C, N, 2]
+            sel = info["radii"] > 0.0  # [C, N]
+            gs_ids = torch.where(sel)[1]  # [nnz]
+            self._state["grad2d"].index_add_(0, gs_ids, grads[sel].norm(dim=-1))
+            self._state["count"].index_add_(
+                0, gs_ids, torch.ones_like(gs_ids, dtype=torch.float32)
+            )
+
+    @torch.no_grad()
+    def _refine_duplicate(self, mask: Tensor):
+        device = mask.device
+        sel = torch.where(mask)[0]
+        for optimizer in self._optimizers:
```

Comment:
do we want some sort of helper function for this? looping through optimizers and adding/pruning/removing is a common pattern that all strategies will need

Reply:
so these lines could then call `self.duplicate_in_optimizer(index_mask)` or something

Reply:
I tried once but didn't end up creating anything elegant. Open for suggestions!

Reply:
[These functions](https://github.com/nerfstudio-project/nerfstudio/blob/27801d2a4786de0f94c280c0fbf27ac717dd0d22/nerfstudio/models/splatfacto.py#L373) in splatfacto try to do something like this, so any subclass could also call dup_in_optim and have the same effect

Reply:
I think the body of the function would need to be different but the general interface of taking in a per-gaussian mask could work for removing/duplicating, or specifying how many to append

Reply:
The nerfstudio way of implementation is mixing the duplication, splitting, and removing operations together so that the remove on optimizer can be done once. But the downside is that it's hard to read and modify. I would like to keep each operation isolated. But yeah maybe there is a way to share some code between them. I just dont have a concrete idea.